File: bindings/typescript/src/internal/permissions.ts
import { decodeBase64 } from './encoding.js';
import type { Permissions } from '../types.js';

export const PERMISSION_GET_INFO = 'getInfo';
export const PERMISSION_CREATE_INVOICE = 'createInvoice';
export const PERMISSION_PAY_INVOICE = 'payInvoice';
export const PERMISSION_CREATE_OFFER = 'createOffer';
export const PERMISSION_GET_OFFER = 'getOffer';
export const PERMISSION_LIST_OFFERS = 'listOffers';
export const PERMISSION_PAY_OFFER = 'payOffer';
export const PERMISSION_LOOKUP_INVOICE = 'lookupInvoice';
export const PERMISSION_LIST_TRANSACTIONS = 'listTransactions';
export const PERMISSION_DECODE = 'decode';
export const PERMISSION_ON_INVOICE_EVENTS = 'onInvoiceEvents';

export const BOLT11_NODE_PERMISSIONS = [
  PERMISSION_GET_INFO,
  PERMISSION_CREATE_INVOICE,
  PERMISSION_PAY_INVOICE,
  PERMISSION_LOOKUP_INVOICE,
  PERMISSION_LIST_TRANSACTIONS,
  PERMISSION_DECODE,
  PERMISSION_ON_INVOICE_EVENTS,
] as const;

export const BOLT12_NODE_PERMISSIONS = [
  PERMISSION_CREATE_OFFER,
  PERMISSION_GET_OFFER,
  PERMISSION_LIST_OFFERS,
  PERMISSION_PAY_OFFER,
] as const;

export const FULL_NODE_PERMISSIONS = [
  ...BOLT11_NODE_PERMISSIONS,
  ...BOLT12_NODE_PERMISSIONS,
] as const;

export const NWC_METHOD_PERMISSIONS = [
  'get_info',
  'get_balance',
  'make_invoice',
  'pay_invoice',
  'lookup_invoice',
  'list_transactions',
] as const;

export const CLN_METHOD_PERMISSIONS = [
  'getinfo',
  'listfunds',
  'invoice',
  'pay',
  'offer',
  'fetchinvoice',
  'listoffers',
  'listinvoices',
  'decode',
] as const;

const STRIKE_SCOPE_PERMISSIONS: Array<{
  permission: string;
  scopes: string[];
}> = [
  {
    permission: PERMISSION_GET_INFO,
    scopes: ['partner.balances.read'],
  },
  {
    permission: PERMISSION_CREATE_INVOICE,
    scopes: ['partner.receive-request.create'],
  },
  {
    permission: PERMISSION_PAY_INVOICE,
    scopes: ['partner.payment-quote.lightning.create', 'partner.payment-quote.execute'],
  },
  {
    permission: PERMISSION_LOOKUP_INVOICE,
    scopes: ['partner.receive-request.read'],
  },
  {
    permission: PERMISSION_LIST_TRANSACTIONS,
    scopes: ['partner.receive-request.read', 'partner.payment.read'],
  },
  {
    permission: PERMISSION_ON_INVOICE_EVENTS,
    scopes: ['partner.receive-request.read'],
  },
];

const BLINK_SCOPE_PERMISSIONS: Array<{
  permission: string;
  scopes: string[];
}> = [
  {
    permission: PERMISSION_GET_INFO,
    scopes: ['read'],
  },
  {
    permission: PERMISSION_CREATE_INVOICE,
    scopes: ['receive'],
  },
  {
    permission: PERMISSION_PAY_INVOICE,
    scopes: ['write'],
  },
  {
    permission: PERMISSION_LOOKUP_INVOICE,
    scopes: ['read'],
  },
  {
    permission: PERMISSION_LIST_TRANSACTIONS,
    scopes: ['read'],
  },
  {
    permission: PERMISSION_ON_INVOICE_EVENTS,
    scopes: ['read'],
  },
];

export function emptyPermissions(): Permissions {
  return {
    getInfo: false,
    createInvoice: false,
    payInvoice: false,
    createOffer: false,
    getOffer: false,
    listOffers: false,
    payOffer: false,
    lookupInvoice: false,
    listTransactions: false,
    decode: false,
    onInvoiceEvents: false,
  };
}

export function isEmptyPermissions(permissions: Permissions): boolean {
  return Object.values(permissions).every((value) => !value);
}

export function normalizePermissionValues(values: Iterable<string | null | undefined>): string[] {
  const seen = new Set<string>();

  for (const value of values) {
    const normalized = value?.trim();
    if (normalized) {
      seen.add(normalized);
    }
  }

  return [...seen].sort((a, b) => a.localeCompare(b));
}

export function normalizePermissions(values: Iterable<string | null | undefined>): Permissions {
  return permissionsFromValues(compactPermissions(values));
}

export function normalizeNwcPermissions(values: Iterable<string | null | undefined>): Permissions {
  const permissions = compactPermissions(values);
  const has = (permission: string) => hasPermission(permissions, permission);

  return {
    ...emptyPermissions(),
    getInfo: has('get_balance'),
    createInvoice: has('make_invoice'),
    payInvoice: has('pay_invoice'),
    lookupInvoice: has('lookup_invoice'),
    listTransactions: has('list_transactions'),
    onInvoiceEvents: has('lookup_invoice'),
  };
}

export function normalizeClnPermissions(values: Iterable<string | null | undefined>): Permissions {
  const permissions = compactPermissions(values);
  const has = (permission: string) => hasPermission(permissions, permission);

  return {
    getInfo: has('getinfo') && has('listfunds'),
    createInvoice: has('invoice'),
    payInvoice: has('pay'),
    createOffer: has('offer'),
    getOffer: has('listoffers'),
    listOffers: has('listoffers'),
    payOffer: has('fetchinvoice') && has('pay'),
    lookupInvoice: has('listinvoices'),
    listTransactions: has('listinvoices'),
    decode: has('decode'),
    onInvoiceEvents: has('listinvoices'),
  };
}

export function normalizeLndPermissions(values: Iterable<string | null | undefined>): Permissions {
  const permissions = compactPermissions(values);
  const has = (permission: string) => hasPermission(permissions, permission);

  return {
    ...emptyPermissions(),
    getInfo:
      (has('/lnrpc.Lightning/GetInfo') && has('/lnrpc.Lightning/ChannelBalance')) ||
      (has('info:read') && has('offchain:read')),
    createInvoice: has('/lnrpc.Lightning/AddInvoice') || has('invoices:write'),
    payInvoice:
      has('/lnrpc.Lightning/SendPaymentSync') ||
      has('/routerrpc.Router/SendPaymentV2') ||
      has('offchain:write'),
    lookupInvoice: has('/lnrpc.Lightning/LookupInvoice') || has('invoices:read'),
    listTransactions:
      has('/lnrpc.Lightning/ListInvoices') ||
      has('/lnrpc.Lightning/ListPayments') ||
      has('invoices:read') ||
      has('offchain:read'),
    decode: has('/lnrpc.Lightning/DecodePayReq') || has('offchain:read'),
    onInvoiceEvents: has('/lnrpc.Lightning/SubscribeInvoices') || has('invoices:read'),
  };
}

export function parseClnRunePermissions(rune: string): Permissions {
  const decoded = decodeBase64Url(rune);
  const text = new TextDecoder().decode(decoded);
  const matches = text.match(/method(?:=|\^)[A-Za-z0-9_.:-]+/g) ?? [];

  if (!matches.length) {
    return normalizeClnPermissions(CLN_METHOD_PERMISSIONS);
  }

  const expanded: string[] = [];
  for (const permission of matches) {
    if (permission.startsWith('method=')) {
      expanded.push(permission.slice('method='.length));
      continue;
    }

    const prefix = permission.slice('method^'.length);
    const matchingMethods = CLN_METHOD_PERMISSIONS.filter((method) => method.startsWith(prefix));
    expanded.push(...(matchingMethods.length ? matchingMethods : [`${prefix}*`]));
  }

  return normalizeClnPermissions(expanded);
}

export function parseLndMacaroonPermissions(bytes: Uint8Array): Permissions {
  const text = new TextDecoder().decode(bytes);
  const matches = text.match(/[a-z][a-z0-9_-]*:(?:read|write|generate)/g) ?? [];
  return normalizeLndPermissions(matches);
}

export function getStrikeOauthPermissions(accessToken: string): Permissions | null {
  const payload = decodeJwtPayload(accessToken);
  if (!payload) {
    return null;
  }

  const scopes = new Set(readScopeValues(payload));
  const permissions = [PERMISSION_DECODE];

  for (const candidate of STRIKE_SCOPE_PERMISSIONS) {
    if (candidate.scopes.every((scope) => scopes.has(scope))) {
      permissions.push(candidate.permission);
    }
  }

  return normalizePermissions(permissions);
}

export function getGaloyTokenPermissions(token: string): Permissions | null {
  const payload = decodeJwtPayload(token);
  if (!payload) {
    return null;
  }

  const scopes = new Set(readScopeValues(payload).map((scope) => scope.toLowerCase()));
  const permissions = [PERMISSION_DECODE];

  for (const candidate of BLINK_SCOPE_PERMISSIONS) {
    if (candidate.scopes.every((scope) => scopes.has(scope))) {
      permissions.push(candidate.permission);
    }
  }

  return normalizePermissions(permissions);
}

/** @deprecated Use `getGaloyTokenPermissions`. */
export const getBlinkTokenPermissions = getGaloyTokenPermissions;

function permissionsFromValues(values: string[]): Permissions {
  const has = (permission: string) => hasPermission(values, permission);

  return {
    getInfo:
      has(PERMISSION_GET_INFO) ||
      has('get_info') ||
      has('getinfo') ||
      has('get_balance') ||
      has('/lnrpc.Lightning/GetInfo') ||
      has('/lnrpc.Lightning/ChannelBalance') ||
      has('info:read'),
    createInvoice:
      has(PERMISSION_CREATE_INVOICE) ||
      has('create_invoice') ||
      has('make_invoice') ||
      has('invoice') ||
      has('/lnrpc.Lightning/AddInvoice') ||
      has('invoices:write'),
    payInvoice:
      has(PERMISSION_PAY_INVOICE) ||
      has('pay_invoice') ||
      has('pay') ||
      has('/lnrpc.Lightning/SendPaymentSync') ||
      has('/routerrpc.Router/SendPaymentV2') ||
      has('offchain:write'),
    createOffer:
      has(PERMISSION_CREATE_OFFER) || has('create_offer') || has('offer') || has('offers:write'),
    getOffer:
      has(PERMISSION_GET_OFFER) || has('get_offer') || has('listoffers') || has('offers:read'),
    listOffers:
      has(PERMISSION_LIST_OFFERS) || has('list_offers') || has('listoffers') || has('offers:read'),
    payOffer:
      has(PERMISSION_PAY_OFFER) ||
      has('pay_offer') ||
      (has('fetchinvoice') && (has('pay') || has(PERMISSION_PAY_INVOICE))),
    lookupInvoice:
      has(PERMISSION_LOOKUP_INVOICE) ||
      has('lookup_invoice') ||
      has('lookup-invoice') ||
      has('listinvoices') ||
      has('/lnrpc.Lightning/LookupInvoice') ||
      has('invoices:read'),
    listTransactions:
      has(PERMISSION_LIST_TRANSACTIONS) ||
      has('list_transactions') ||
      has('listinvoices') ||
      has('/lnrpc.Lightning/ListInvoices') ||
      has('/lnrpc.Lightning/ListPayments') ||
      has('invoices:read') ||
      has('offchain:read'),
    decode: has(PERMISSION_DECODE) || has('/lnrpc.Lightning/DecodePayReq') || has('address:read'),
    onInvoiceEvents:
      has(PERMISSION_ON_INVOICE_EVENTS) ||
      has('on_invoice_events') ||
      has('lookup_invoice') ||
      has('listinvoices') ||
      has('/lnrpc.Lightning/SubscribeInvoices') ||
      has('invoices:read'),
  };
}

function compactPermissions(values: Iterable<string | null | undefined>): string[] {
  return [...values].flatMap((value) => {
    const normalized = value?.trim();
    return normalized ? [normalized] : [];
  });
}

function hasPermission(values: string[], permission: string): boolean {
  return values.some((value) => value.toLowerCase() === permission.toLowerCase());
}

function decodeJwtPayload(accessToken: string): Record<string, unknown> | null {
  const parts = accessToken.split('.');
  const payload = parts[1];
  if (parts.length < 3 || !payload) {
    return null;
  }

  try {
    return JSON.parse(new TextDecoder().decode(decodeBase64Url(payload))) as Record<
      string,
      unknown
    >;
  } catch {
    return null;
  }
}

function readScopeValues(payload: Record<string, unknown>): string[] {
  const values = [payload.scope, payload.scp, payload.scopes];
  const scopes: string[] = [];

  for (const value of values) {
    if (typeof value === 'string') {
      scopes.push(...value.split(/\s+/));
      continue;
    }

    if (Array.isArray(value)) {
      scopes.push(...value.filter((item): item is string => typeof item === 'string'));
    }
  }

  return normalizePermissionValues(scopes);
}

function decodeBase64Url(input: string): Uint8Array {
  const normalized = input.trim().replace(/-/g, '+').replace(/_/g, '/');
  const padding = (4 - (normalized.length % 4 || 4)) % 4;
  return decodeBase64(normalized + '='.repeat(padding));
}

File: crates/lni/permissions.rs
use regex::Regex;

const NWC_METHOD_PERMISSIONS: &[&str] = &[
    "get_info",
    "get_balance",
    "make_invoice",
    "pay_invoice",
    "lookup_invoice",
    "list_transactions",
];

const CLN_METHOD_PERMISSIONS: &[&str] = &[
    "getinfo",
    "listfunds",
    "invoice",
    "pay",
    "offer",
    "fetchinvoice",
    "listoffers",
    "listinvoices",
    "decode",
];

const STRIKE_SCOPE_PERMISSIONS: &[(&str, &[&str])] = &[
    ("get_info", &["partner.balances.read"]),
    ("create_invoice", &["partner.receive-request.create"]),
    (
        "pay_invoice",
        &[
            "partner.payment-quote.lightning.create",
            "partner.payment-quote.execute",
        ],
    ),
    ("lookup_invoice", &["partner.receive-request.read"]),
    (
        "list_transactions",
        &["partner.receive-request.read", "partner.payment.read"],
    ),
    ("on_invoice_events", &["partner.receive-request.read"]),
];

const BLINK_SCOPE_PERMISSIONS: &[(&str, &[&str])] = &[
    ("get_info", &["read"]),
    ("create_invoice", &["receive"]),
    ("pay_invoice", &["write"]),
    ("lookup_invoice", &["read"]),
    ("list_transactions", &["read"]),
    ("on_invoice_events", &["read"]),
];

pub fn nwc_method_permissions() -> Vec<String> {
    normalize_permissions(
        NWC_METHOD_PERMISSIONS
            .iter()
            .map(|permission| (*permission).to_string()),
    )
}

pub fn normalize_permissions<I>(permissions: I) -> Vec<String>
where
    I: IntoIterator<Item = String>,
{
    let mut values: Vec<String> = permissions
        .into_iter()
        .map(|permission| permission.trim().to_string())
        .filter(|permission| !permission.is_empty())
        .collect();
    values.sort();
    values.dedup();
    values
}

pub fn parse_cln_rune_permissions(rune: &str) -> Vec<String> {
    let decoded = match base64::decode_config(pad_base64_url(rune), base64::URL_SAFE) {
        Ok(bytes) => bytes,
        Err(_) => {
            return normalize_permissions(
                CLN_METHOD_PERMISSIONS
                    .iter()
                    .map(|permission| (*permission).to_string()),
            )
        }
    };
    let text = String::from_utf8_lossy(&decoded);
    let matcher = Regex::new(r"method(?:=|\^)[A-Za-z0-9_.:-]+").expect("valid rune method regex");
    let matches: Vec<&str> = matcher
        .find_iter(&text)
        .map(|match_| match_.as_str())
        .collect();

    if matches.is_empty() {
        return normalize_permissions(
            CLN_METHOD_PERMISSIONS
                .iter()
                .map(|permission| (*permission).to_string()),
        );
    }

    let mut expanded = Vec::new();
    for permission in matches {
        if let Some(method) = permission.strip_prefix("method=") {
            expanded.push(method.to_string());
            continue;
        }

        if let Some(prefix) = permission.strip_prefix("method^") {
            let mut matched = false;
            for method in CLN_METHOD_PERMISSIONS {
                if method.starts_with(prefix) {
                    expanded.push((*method).to_string());
                    matched = true;
                }
            }
            if !matched {
                expanded.push(format!("{prefix}*"));
            }
        }
    }

    normalize_permissions(expanded)
}

pub fn parse_lnd_macaroon_permissions(bytes: &[u8]) -> Vec<String> {
    let text = String::from_utf8_lossy(bytes);
    let matcher = Regex::new(r"[a-z][a-z0-9_-]*:(?:read|write|generate)")
        .expect("valid lnd macaroon permission regex");
    normalize_permissions(
        matcher
            .find_iter(&text)
            .map(|match_| match_.as_str().to_string()),
    )
}

pub fn parse_strike_oauth_permissions(access_token: &str) -> Option<Vec<String>> {
    let payload = decode_jwt_payload(access_token)?;
    let scopes = read_scope_values(&payload);
    let mut permissions = vec!["decode".to_string()];

    for (permission, required_scopes) in STRIKE_SCOPE_PERMISSIONS {
        if required_scopes.iter().all(|scope| scopes.iter().any(|item| item == scope)) {
            permissions.push((*permission).to_string());
        }
    }

    Some(normalize_permissions(permissions))
}

pub fn parse_blink_token_permissions(token: &str) -> Option<Vec<String>> {
    let payload = decode_jwt_payload(token)?;
    let scopes: Vec<String> = read_scope_values(&payload)
        .into_iter()
        .map(|scope| scope.to_lowercase())
        .collect();
    let mut permissions = vec!["decode".to_string()];

    for (permission, required_scopes) in BLINK_SCOPE_PERMISSIONS {
        if required_scopes.iter().all(|scope| scopes.iter().any(|item| item == scope)) {
            permissions.push((*permission).to_string());
        }
    }

    Some(normalize_permissions(permissions))
}

fn decode_jwt_payload(access_token: &str) -> Option<serde_json::Value> {
    let mut parts = access_token.split('.');
    let _header = parts.next()?;
    let payload = parts.next()?;
    let _signature = parts.next()?;
    let decoded = base64::decode_config(pad_base64_url(payload), base64::URL_SAFE).ok()?;
    serde_json::from_slice(&decoded).ok()
}

fn read_scope_values(payload: &serde_json::Value) -> Vec<String> {
    let mut scopes = Vec::new();

    for key in ["scope", "scp", "scopes"] {
        match payload.get(key) {
            Some(serde_json::Value::String(value)) => {
                scopes.extend(value.split_whitespace().map(|scope| scope.to_string()));
            }
            Some(serde_json::Value::Array(values)) => {
                scopes.extend(values.iter().filter_map(|value| value.as_str().map(|scope| scope.to_string())));
            }
            _ => {}
        }
    }

    normalize_permissions(scopes)
}

fn pad_base64_url(input: &str) -> String {
    let normalized = input.trim();
    let padding = (4 - (normalized.len() % 4)) % 4;
    format!("{}{}", normalized, "=".repeat(padding))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn expands_cln_method_prefix_restrictions() {
        let rune = base64::encode_config("unique-id&method^list|method=getinfo", base64::URL_SAFE);

        assert_eq!(
            parse_cln_rune_permissions(&rune),
            vec![
                "getinfo".to_string(),
                "listfunds".to_string(),
                "listinvoices".to_string(),
                "listoffers".to_string(),
            ]
        );
    }

    #[test]
    fn parses_lnd_permission_pairs_from_macaroon_bytes() {
        assert_eq!(
            parse_lnd_macaroon_permissions(b"info:read invoices:write offchain:read"),
            vec![
                "info:read".to_string(),
                "invoices:write".to_string(),
                "offchain:read".to_string(),
            ]
        );
    }

    #[test]
    fn maps_strike_oauth_scopes_to_lni_permissions() {
        let header = base64_url_json(&serde_json::json!({ "alg": "none" }));
        let payload = base64_url_json(&serde_json::json!({
            "scope": "openid partner.balances.read partner.receive-request.create partner.receive-request.read partner.payment-quote.lightning.create partner.payment-quote.execute"
        }));
        let access_token = format!("{header}.{payload}.signature");

        assert_eq!(
            parse_strike_oauth_permissions(&access_token),
            Some(vec![
                "create_invoice".to_string(),
                "decode".to_string(),
                "get_info".to_string(),
                "lookup_invoice".to_string(),
                "on_invoice_events".to_string(),
                "pay_invoice".to_string(),
            ])
        );
    }

    #[test]
    fn rejects_opaque_strike_api_keys() {
        assert_eq!(parse_strike_oauth_permissions("sk_test_opaque"), None);
    }

    #[test]
    fn maps_blink_scopes_to_lni_permissions() {
        let header = base64_url_json(&serde_json::json!({ "alg": "none" }));
        let payload = base64_url_json(&serde_json::json!({ "scope": "Read Receive Write" }));
        let token = format!("{header}.{payload}.signature");

        assert_eq!(
            parse_blink_token_permissions(&token),
            Some(vec![
                "create_invoice".to_string(),
                "decode".to_string(),
                "get_info".to_string(),
                "list_transactions".to_string(),
                "lookup_invoice".to_string(),
                "on_invoice_events".to_string(),
                "pay_invoice".to_string(),
            ])
        );
    }

    #[test]
    fn rejects_opaque_blink_api_keys() {
        assert_eq!(parse_blink_token_permissions("blink_opaque"), None);
    }

    fn base64_url_json(value: &serde_json::Value) -> String {
        base64::encode_config(value.to_string(), base64::URL_SAFE).trim_end_matches('=').to_string()
    }
}
